File: src/main/java/ar/edu/unq/spring/service/impl/InventarioServiceImp.java
package ar.edu.unq.spring.service.impl;

import ar.edu.unq.spring.modelo.Item;
import ar.edu.unq.spring.modelo.Personaje;
import ar.edu.unq.spring.persistence.ItemDAO;
import ar.edu.unq.spring.persistence.PersonajeDAO;
import ar.edu.unq.spring.service.interfaces.InventarioService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collection;
import java.util.Optional;

@Service
@Transactional
public class InventarioServiceImp implements InventarioService {

    @Autowired
    private PersonajeDAO personajeDAO;

    @Autowired
    private ItemDAO itemDAO;

    @Override
    public Item getItem(Long itemId) {
        return itemDAO.findById(itemId).orElse(null);
    }

    @Override
    public Collection<Item> allItems() {
        return itemDAO.findAll();
    }

    @Override
    public Item heaviestItem() {
        return itemDAO.findTopByOrderByPesoDesc();
    }

    @Override
    public Item guardarItem(Item item) {
        return itemDAO.save(item);
    }

    @Override
    public void recoger(Long personajeId, Long itemId) {
        Optional<Personaje> personaje = personajeDAO.findById(personajeId);
        Optional<Item> item = itemDAO.findById(itemId);
        if (personaje.isPresent() && item.isPresent()) {
            personaje.get().recoger(item.get());
            personajeDAO.save(personaje.get());
        }
    }

    @Override
    public Collection<Item> getMasPesados(int peso) {
        return itemDAO.getMasPesados(peso);
    }

    @Override
    public Collection<Item> getItemsPersonajesDebiles(int vida) {
        return itemDAO.getItemsDePersonajesDebiles(vida);
    }

    @Override
    public void clearAll() {
        itemDAO.deleteAll();
    }

    @Override
    public void deleteItem(Item item) {
        itemDAO.delete(item);
    }
}

File: src/main/kotlin/ar/edu/unq/spring/service/impl/InventarioServiceImp.kt
package ar.edu.unq.spring.service.impl

import ar.edu.unq.spring.modelo.Item
import ar.edu.unq.spring.persistence.ItemDAO
import ar.edu.unq.spring.persistence.PersonajeDAO
import ar.edu.unq.spring.service.interfaces.InventarioService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class InventarioServiceImp() : InventarioService {

    @Autowired private lateinit var personajeDAO: PersonajeDAO
    @Autowired private lateinit var itemDAO: ItemDAO
    override fun getItem(itemId: Long): Item {
        return itemDAO.findById(itemId).get()
    }

    override fun allItems(): Collection<Item> {
        return itemDAO.findAll().toMutableSet()
    }

    override fun heaviestItem(): Item {
        return itemDAO.findTopByOrderByPesoDesc()
    }

    override fun guardarItem(item: Item): Item {
       return itemDAO.save(item)
    }

    override fun recoger(personajeId: Long, itemId: Long) {
        val personaje = personajeDAO.findByIdOrNull(personajeId)
        val item = itemDAO.findByIdOrNull(itemId)
        item?.let { i -> personaje?.recoger(i) }
        personaje?.let { p -> personajeDAO.save(p) }
    }

    override fun getMasPesados(peso: Int): Collection<Item> {
        return itemDAO.getMasPesados(peso)
    }

    override fun getItemsPersonajesDebiles(vida: Int): Collection<Item> {
        return itemDAO.getItemsDePersonajesDebiles(vida)
    }

    override fun clearAll() {
        itemDAO.deleteAll()
    }

    override fun deleteItem(item: Item) {
        itemDAO.delete(item)
    }

}
